project/CopyTruffleJAR.scala

```diff
@@ -0,0 +1,124 @@
+        baseDirectory.value,
+        (Compile / update).value,
+        log
+      )
+    ) {
+      log.info("Truffle JARs have been updated.")
+      System.err.println(
+        "You have to restart the sbt JVM for the changes to take effect."
+      )
+      System.out.flush()
+      System.err.flush()
+      System.exit(0)
+    } else {
+      log.info("Truffle JARs are up to date.")
+    }
+  }
+
+  /**
+    * This task should be added as a dependency of compileInputs in the runtime
+    * subproject. It ensures that the compilation will not proceed unless the
+    * JARs have not been bootstrapped. If the JARs were out of date, they are
+    * updated within this task, so bootstrap does not have to be re-run.
+    * However, for the same reasons as for [[bootstrapJARs]], the sbt process is
+    * terminated, because a restart is required.
+    */
+  lazy val preCompileTask = Def.task {
+    val log = streams.value.log
+    if (
+      ensureTruffleJARUpToDate(
+        baseDirectory.value,
+        (Compile / update).value,
+        log
+      )
+    ) {
+      log.error(
+        "JARs that have to be loaded by the sbt JVM at startup have been" +
+        " modified or did not exist.\n" +
+        "Did you run bootstrap?"
+      )
+      log.warn(
+        "Bootstrap has been triggered automatically, but the sbt JVM must be" +
+        " restarted to apply the changes, so the compilation had to be stopped."
+      )
+      log.warn(
+        "To avoid disrupting compilation, remember to run bootstrap when " +
+        "setting up the project and after each version change of Graal."
+      )
+      System.err.println(
+        "The sbt JVM has to be restarted to apply the changes.\n" +
+        "Please re-launch sbt and re-run the last command."
+      )
+      System.out.flush()
+      System.err.flush()
+      System.exit(0)
+    }
+  }
+
+  /**
+    * Checks the Truffle JARs and updates them if necessary.
+    *
+    * @param libraryUpdates the value of Compile / updates
+    * @return true if an update has been performed and the JVM needs a restart
+    */
+  private def ensureTruffleJARUpToDate(
+    baseDirectory: File,
+    libraryUpdates: UpdateReport,
+    log: ManagedLogger
+  ): Boolean = {
+    var truffleInstancesFound = 0
+    var restartRequired       = false
+    libraryUpdates.allFiles.foreach { f =>
+      if (f.getName.contains("truffle-api")) {
+        truffleInstancesFound += 1
+        val dest = baseDirectory / "build-cache" / "truffle-api.jar"
+        val needsUpdate = if (!dest.exists()) {
+          log.debug("truffle-api.jar does not exist in target/")
+          true
+        } else if (dest.lastModified() < f.lastModified()) {
+          log.debug("truffle-api.jar in target/ is out of date")
+          true
+        } else false
+
+        if (needsUpdate) {
+          IO.copyFile(f, dest)
+          restartRequired = true
+        }
+      }
+    }
+
+    if (truffleInstancesFound == 0) {
```

Comment:
These error messages are completely unhelpful, let me know what to do when things happen. For the first case – what's the root cause? Was bootstrap not called? Tell me to run it. For the second case – "please make sure everything is configured properly." – yeah ok, what does that mean? Do I rerun bootstrap? Do I run some other command?

Reply:
These two should never be triggered actually - I just added them as a safeguard.
The first one would be triggered if Truffle is not found in the dependencies (perhaps the dependency was renamed?). 

The second one would happen if two dependencies matched the current way of 'finding' the right Turffle API dependency and so the task was unable to do its job properly (as it doesn't know which one to pick).

Both require modifying build.sbt or this task and may only be triggered if build.sbt is modified.

Reply:
Maybe I should add `If dependencies have been changed in build.sbt, make sure the algorithm locating truffle-api.jar in project/CopyTruffleJAR.scala has been updated accordingly.` to both of these error messages?

Reply:
I see, so only the person doing the bumping should ever see these? In this case I would go with the standard `please report this as a bug` – I bet someone will see this issue randomly anyway, so it's probably good to let them know they shouldn't be seeing this.